### DISCUSSION TITLE
New Writer/Generator/Creator API

### DISCUSSION THREAD

**Discussion by axxel** [zxing-cpp/zxing-cpp#739] [General]
Given that the "discussion" in #332 is lengthy and mostly outdated and also not in form of a github discussion, I summarize my current state of mind here:

* the c++ implementations of the barcode writers in this repo will (in the long run) be completely removed and replaced by [libzint](https://github.com/zint/zint).
* the `Barcode` class will be the central point of both reading and generating barcodes and can also be written/rendered into an image/svg.
* there will be consistent C++ API for both reading and generating barcodes that (almost) completely hides the fact that the `zint` c-library is used behind the scenes.
* there will be a C-API that closely maps to the C++-API for both applications, completely independent of the existing `zint` API.
* there will be wrappers, partly based on that C-API, that model the C++-API more or less, depending on what feels 'native' in the respective language. Every wrapper should feel as if the library was written in that language natively.
* the C++ "wrapper" around zint will not strive to expose all bells and whistles / features of the `zint` library, e.g. only the symbologies currently known to `zxing-cpp` will be exposed; the segmented ECI block feature will likely not be (maybe later if there is demand); writing directly to png-files will not be, the option to pass non-utf8 text might not be supported, etc.
* there might be a build option to choose either the old backend or `libzint` depending on the tradeoffs of the two solutions.

Now is the time to communicate your needs, use cases and wishes with regards to this API. Especially, if it is something that is in conflict with my intentions stated above. The specific "mechanics" of a potential 'fluent interface' API are discussed [here](https://github.com/zxing-cpp/zxing-cpp/discussions/724#discussioncomment-8567239).

Update: to try out the new API with the current HEAD, `zxing-cpp` needs to be configured with `ZXING_WRITERS=NEW` and `ZXING_EXPERIMENTAL_API=ON`. Working sample code can be found [here](https://github.com/zxing-cpp/zxing-cpp/blob/e2da6952ca475b1c4d15b94306996de0a8b68e81/example/ZXingWriter.cpp#L171-L191).

**Comment by axxel** [comment]
I added 2 changesets:

* 3a79c38e65460d8ea31273d0d9cdc47a0be63ffc: add a new experimental 'create' and 'write' (next to 'read') API along the lines of what is discussed above
* 0ee68dfc3224cfcdb623c693c2b8fa40d8b2daa0: add libzint based backend for new API, needs `ZXING_USE_ZINT=ON` cmake option

The shortest possible use case in c++ with the current function based API (the 'fluent interface' is still on the table / in the making) looks like this:
```c++
auto svg = WriteBarcodeToSVG(CreateBarcodeFromText("hello", BarcodeFormat::QRCode));
```

A more realistic scenario with an image instead of an SVG string would be:

```c++
auto cOpts = CreatorOptions(BarcodeFormat::QRCode).ecLevel("M");
auto bcode = CreateBarcodeFromText("hello", cOpts);
auto wOpts = WriterOptions().sizeHint(400).withQuietZones(true).withHRT(true);
auto image = WriteBarcodeToImage(barcode, wOpts);
// write_to_file(image.data(), image.width(), image.height(), image.format());
```

Looking for feedback...

**Reply by axxel** [reply]
The exact list of exposed creator/writer options is not yet finalized. The current list can be seen in the source code (see linked changesets above). If someone needs a specific feature, let me know. There are a number of fundamental incompatibilities between the old `MultiFormatWriter` and the new API:

* explicit `width + height` becomes either 
 a) `sizeHint`: a size in pixel for the longer side of the symbol that is approximately returned, the exact size may differ and depends on heuristics that are there to prevent issues like #495 
 b) `scale`: this is basically the module size - maybe I should actually call it `moduleSize`?
* `ecLevel` is now a string that can contain 1 of 3 types of information a) a letter M,L,H,Q for QRCodes b) a percentage like `50%` that works across all symbologies and will internally pick the closes available value and c) an integer that has a symbologies dependent meaning, exactly as in the `zint_symbol->option_1` API.
* I see not much value in support the `margin` property that simply adds a bunch of white pixels around the image. That can always be done in client code.
* I strongly intend to not expose the `CharcterSet` property.

**Reply by axxel** [reply]
38bf41654e52ab1ba33ba48c77c5c5a65d768f97: Adds C-API for the creator/writer

**Reply by axxel** [reply]
4710c4c51ad9146fcb862e7b7af097b1e07cacba: `BUILD_WRITER` has now 5 possible states: `OFF/ON/OLD/NEW/BOTH`. `OLD` is the exiting backend, `NEW` is the zint-based one, `BOTH` provides both (keeps `MultiFormatWriter` available), `ON` currently defaults to `OLD`.
2d29e0eb9a845b4d89503f599f248a16ed051096: Adds a `ZXING_USE_BUNDLED_ZINT` cmake build option that compiles (a subset of) libzint, pulled in via git submodule, statically into `zxing-cpp`. This will not be of interest for linux distributions, who can simply depend on system libzint (but require at least the yet unreleased version 2.13). But it will certainly be the way to go for the wrappers. The option is not used anywhere yet and only has any effect if the `BUILD_WRITER` option is set to either `NEW` or `BOTH`.

**Reply by axxel** [reply]
9a17b97e248e312f3433c3c26c86dfe1bda2304c: adds a writer API to the Rust wrapper

**Reply by Sec-ant** [reply]
There're both `ecLevel` and `eccLevel` these 2 names in the codebase, should them be unified?

**Reply by axxel** [reply]
That is exactly what I did, see [here](https://github.com/zxing-cpp/zxing-cpp/blob/b58682b90ff082cee8d946c73f8852574478fb09/core/src/Barcode.h#L95).

**Reply by axxel** [reply]
https://github.com/zxing-cpp/zxing-cpp/commit/66accce20102030f9feb994687ba293afdc96eeb: adds writer API to .NET wrapper

**Reply by farfromrefug** [reply]
@axxel i am trying your code with `WriterOptions` but i get an error `Incomplete type 'ZXing::WriterOptions' `

Any idea what might be going on? I am using c++ 20

EDIT: actually seems to be an IDE thing as it is building correctly

**Comment by Sec-ant** [comment]
Hi @axxel I don't have suggestions or objections whatsoever, just reply to let you know I'm following this :)

Though I'd like to kindly request a draft of the API structure in your mind, like what are the fields and values of the r/w Options and Barcode object, what are the names we plan to use, etc. Right now this kind of information is spreaded across several issues and discussions, which takes some time to look forth and behind. It will make it easier for wrapper developers to make an early move and decide how to properly deprecate some properties and refactor their codebases.

Please ignore my request if this adds unnecessary burden to you. I appreciate all your work put on this.

**Comment by axxel** [comment]
@vkrause you mentioned a while back that there might be some interest from your side to use the writer/creator capabilities of zxing-cpp instead of using separate libraries for individual symbologies. Is that still something of interest? Can you have look at the current API proposal (e.g. see [above](https://github.com/zxing-cpp/zxing-cpp/discussions/739#discussioncomment-8660387) or [here](https://github.com/zxing-cpp/zxing-cpp/blob/5f1d84f0a9aab6bbe9317fade54a8c6151573ba0/example/ZXingWriter.cpp#L162-L166)) and tell me if that would fit your bill?

Is there something you specifically need or missed with the old writer support?

Do you have a use for the `WriteBarcodeToSVG()` function including the `withHRT` (human readable text) option that the libzint backend brings to the table?

**Reply by vkrause** [reply]
That's still of interest, although I haven't pursued this further yet as the future of the writer support sounded a bit uncertain back then. The exact API style doesn't really matter, as this will be an implementation detail behind our own API anyway (as we need to integrate this with Qt rendering). So as long as we can input text and raw binary data (for symbologies supporting that), specify the error correction level (and maybe the PDF417 aspect ratio), and get some form of bit matrix out we are fine. Everything I have seen proposed would work for us.

For the output we don't need anything but the barcode bit matrix itself, ie. no SVG, no image writing, no human readable text, no scaling, no expanding 1D barcodes to a certain height, not even quiet zones.

**Comment by farfromrefug** [comment]
@axxel using the new API (generating SVG string), if the text passed to `CreateBarcodeFromText` is not a "valid" text for the format it seems to go in an endless loop.
Is there a way to test if a string is valid for a format?
Could it be that you dont use `ZBarcode_Encode_and_Print`  as explained [here](https://www.zint.org.uk/manual/chapter/5)? 
EDIT: i managed to catch the std exception. Still wondering if i could "check" a string against a format without the need to catch the exception

**Reply by axxel** [reply]
What do you mean with "endless loop"? You mean the `CreateBarcodeFromText` function never returns? In that case, please open an issue and provide a sample text input that shows the problem.

There is no other way to "check" the validity of a string other than calling the function and catching and evaluating any potentially thrown exception. Exceptions are the means of transporting error information in this API.

**Reply by farfromrefug** [reply]
@axxel i thought it was an endless loop because my cpu was going crazy at that time. But now seeing i managed to simply catch the exception, i think it must have been something else. Sorry.
Ok for the check i ll keep on doing it with the exception handling

**Comment by farfromrefug** [comment]
@axxel i forgot to mention that the new API works great. I tested image generation and svg. I am using svg right now and it is great!
One side note on that. My app have 2 flavors, one with qrcode and one without. Comparing the 2 i can see that zxing adds 2 Mb to my so lib (on android)  in release mode. Is that coming from zint? (did not have the time to test older zxing versions not using zint).
Seems quite big but not sure if we can do anything about it

**Reply by axxel** [reply]
No idea why I missed answering this question... Good thing: things changed in your favor. There are new cmake options available to selectively enable/disable individual symbologies (like `-DZXING_ENABLE_PDF417=OFF`). Depending on what you need, this can lower the binary size for you. Also make sure to set `ZXING_WRITERS=NEW`, which disables the inclusion of the old `zxing-cpp` writer code. Please let me know what you find, if you are still interested in this subject.

**Comment by axxel** [comment]
In preparation for the next release, I went over the API design again and did some changes, here is an example:
```c++
auto cOpts = CreatorOptions(BarcodeFormat::QRCode. "ecLevel:M");
auto bcode = CreateBarcodeFromText("hello", cOpts);
auto wOpts = WriterOptions().sizeHint(400).addQuietZones(true).addHRT(true);
auto image = WriteBarcodeToImage(barcode, wOpts);
// write_to_file(image.data(), image.width(), image.height(), image.format());
```

Most notable, `CreatorOptions` has now an `options` string property that gets passed a key-value pair string (e.g. JSON dict) with barcode specific options, like `ecLevel`, `forceSquare`, `readerInit`, `version` and a bunch of others.

`sizeHint` is still up for debate. There is currently also `scale` which sets the module size, while `sizeHint` specifies the approximate target size (meaning the module size is chosen to match that as close as possible). Having two options that influence the same thing and need to be prioritized if given both is not to my liking. I'm inclined to merge them and choose the interpretation of the number either based on value (like `<=10` -> module size, otherwise target width) or sign (like negative numbers mean one thing while positive ones the other). Both is not ideal either. Comments welcome.

Also related is [this question](https://github.com/zxing-cpp/zxing-cpp/discussions/706#discussioncomment-15289907).

**Comment by axxel** [comment]
@gitlost is there anything from the public zint API currently not available via the `CreatorOptions` that you deem important to add before releasing 3.0?

**Comment by gitlost** [comment]
Looking at the diff between the "diagnostics2" branch and master, the only
important one is "tryCode39ExtendedMode", which I only added after toying
with the alternative of adding a `Code39Ext` variant to `BarcodeFormat`.

I think having one or the other would be good, to signal the use of
`BARCODE_EXCODE39` vs. `BARCODE_CODE39` on calling zint, rather than using
the `iscntrl` workaround (which actually only sets the SI modifier
currently in master, and not the zint `symbology`, so fails).

**Reply by axxel** [reply]
Am I correct in assuming that the only difference between `BARCODE_CODE39` (a) and `BARCODE_EXCODE39` (b) is whether zint complains when encountering a `full ascii` character (a) or translates any of those into 2 character standard content (b). So simply using (b) all the time will produce correct symbols, no matter what is passed to `CreateBarcodeFrom...`?

**Comment by gitlost** [comment]
No, the 4 Code39 characters "$", "/", "+" and "%" are mapped to
double-character escape sequences (they're the ones used as the initial
escape character).
